// angular-interceptors/esm5/src/ensure-https.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { Injectable } from '@angular/core';
var EnsureHttpsInterceptor = /** @class */ (function () {
    function EnsureHttpsInterceptor() {
        this.cache = new Map();
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    EnsureHttpsInterceptor.prototype.intercept = function (req, next) {
        return next.handle(req.clone({ url: req.url.replace('http://', 'https://') }));
    };
    return EnsureHttpsInterceptor;
}());
export { EnsureHttpsInterceptor };
EnsureHttpsInterceptor.decorators = [
    { type: Injectable },
];
/** @nocollapse */
EnsureHttpsInterceptor.ctorParameters = function () { return []; };
function EnsureHttpsInterceptor_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    EnsureHttpsInterceptor.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    EnsureHttpsInterceptor.ctorParameters;
    /** @type {?} */
    EnsureHttpsInterceptor.prototype.cache;
}
export var /** @type {?} */ ENSURE_HTTPS_INTERCEPTOR_PROVIDER = {
    provide: HTTP_INTERCEPTORS,
    useClass: EnsureHttpsInterceptor,
    multi: true
};
