// angular-interceptors/esm5/src/prefix-url.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { Inject, Injectable, InjectionToken } from '@angular/core';
export var /** @type {?} */ PREFIX = new InjectionToken('prefix');
var PrefixUrlInterceptor = /** @class */ (function () {
    /**
     * @param {?} prefix
     */
    function PrefixUrlInterceptor(prefix) {
        this.prefix = prefix;
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    PrefixUrlInterceptor.prototype.intercept = function (req, next) {
        return next.handle(req.clone({ url: this.prefix + req.url }));
    };
    return PrefixUrlInterceptor;
}());
export { PrefixUrlInterceptor };
PrefixUrlInterceptor.decorators = [
    { type: Injectable },
];
/** @nocollapse */
PrefixUrlInterceptor.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: Inject, args: [PREFIX,] },] },
]; };
function PrefixUrlInterceptor_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    PrefixUrlInterceptor.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    PrefixUrlInterceptor.ctorParameters;
    /** @type {?} */
    PrefixUrlInterceptor.prototype.prefix;
}
export var /** @type {?} */ PREFIX_URL_INTERCEPTOR_PROVIDER = {
    provide: HTTP_INTERCEPTORS,
    useClass: PrefixUrlInterceptor,
    multi: true
};
