// angular-interceptors/esm2015/src/cache.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { Inject, Injectable, InjectionToken, Optional } from '@angular/core';
import { interval, throwError } from 'rxjs';
import { catchError, finalize, publishReplay, refCount, take } from 'rxjs/operators';
export const /** @type {?} */ MAX_AGE_MS = new InjectionToken('maxAgeMs');
export class CacheInterceptor {
    /**
     * @param {?} maxAgeMs
     */
    constructor(maxAgeMs) {
        this.maxAgeMs = maxAgeMs;
        this.cache = new Map();
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    intercept(req, next) {
        if (!this.isCachable(req)) {
            return next.handle(req);
        }
        const /** @type {?} */ cachedResponse = this.cache.get(req.urlWithParams);
        if (cachedResponse) {
            return cachedResponse;
        }
        const /** @type {?} */ obs$ = next.handle(req).pipe(finalize(() => interval(this.maxAgeMs)
            .pipe(take(1))
            .subscribe(() => this.cache.delete(req.urlWithParams))), publishReplay(), refCount(), catchError(err => throwError(err)));
        this.cache.set(req.urlWithParams, obs$);
        return obs$;
    }
    /**
     * @param {?} req
     * @return {?}
     */
    isCachable(req) {
        return req.method === 'GET' && !req.headers.get('disable-cache');
    }
}
CacheInterceptor.decorators = [
    { type: Injectable },
];
/** @nocollapse */
CacheInterceptor.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [MAX_AGE_MS,] }, { type: Optional },] },
];
function CacheInterceptor_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    CacheInterceptor.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    CacheInterceptor.ctorParameters;
    /** @type {?} */
    CacheInterceptor.prototype.cache;
    /** @type {?} */
    CacheInterceptor.prototype.maxAgeMs;
}
export const /** @type {?} */ CACHE_INTERCEPTOR_PROVIDER = {
    provide: HTTP_INTERCEPTORS,
    useClass: CacheInterceptor,
    multi: true
};
