// angular-interceptors/esm2015/src/prefix-url.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { PREFIX, PREFIX_URL_INTERCEPTOR_PROVIDER } from './prefix-url.interceptor';
export { PREFIX } from './prefix-url.interceptor';
export class PrefixUrlInterceptorModule {
    /**
     * Prefix HTTP request urls.
     * @param {?} prefix
     * @return {?}
     */
    static forRoot(prefix) {
        return {
            ngModule: PrefixUrlInterceptorModule,
            providers: [PREFIX_URL_INTERCEPTOR_PROVIDER, { provide: PREFIX, useValue: prefix }]
        };
    }
}
PrefixUrlInterceptorModule.decorators = [
    { type: NgModule },
];
function PrefixUrlInterceptorModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    PrefixUrlInterceptorModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    PrefixUrlInterceptorModule.ctorParameters;
}
